### PR TITLE
[willow] new extractor

### PR DESCRIPTION
**Pull request by aarubui**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Extractor for willow.tv, which has both free and subscriber-only content. Only support for free videos was added as I am not a paid subscriber.

### CODE REVIEW

File: supportedsites.md [hunk 1]
```diff
@@ -1311,6 +1311,7 @@
  - **WeiboMobile**
  - **WeiqiTV**: WQTV
  - **whowatch**
+ - **Willow**
```
Comment (pukkandan) [inline supportedsites.md:1314]: Remove - this will be auto-generated at release

File: yt_dlp/extractor/willow.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        video_data = self._parse_json(self._html_search_regex(
+            r'var\s+data_js\s*=\s*JSON\.parse\(\'(.+)\'\)', webpage,
+            'data_js'), video_id)
+        videos = video_data.get('trending_videos') or []
```
Comment (pukkandan) [inline yt_dlp/extractor/willow.py:41]: why trending_videos? Is this first one in this list always guarenteed to be the main video?

Comment (aarubui) [reply yt_dlp/extractor/willow.py:41]: Despite the name, the first item in `trending_videos` is the main video, the remaining are shown as recommendations beside it and have no `secureurl` property. I made it loop through and check for `secureurl` to be flexible but it's always `video_data['trending_videos'][0]`.

File: yt_dlp/extractor/willow.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                        'twitter:title', webpage, default=None),
+                    'thumbnail': v.get('yt_thumb_url') or self._html_search_meta(
+                        'twitter:image', webpage, default=None),
+                    'duration': v.get('duration_seconds'),
+                    'timestamp': v.get('created_date'),
+                    'location': v.get('venue'),
+                    'series': v.get('series_name'),
+                }
+
+        raise ExtractorError('No videos found')
```
Comment (pukkandan) [inline yt_dlp/extractor/willow.py:64]:
```suggestion

        video = next((v for v in video_data.get('trending_videos') or []
                      if v.get('secureurl')), None)
        if not video:
            raise ExtractorError('No videos found')

        formats = self._extract_m3u8_formats(video['secureurl'], video_id, 'mp4')
        self._sort_formats(formats)

        return {
            'id': str(video.get('content_id')),
            'display_id': video.get('video_slug'),
            'title': video.get('video_name') or self._html_search_meta('twitter:title', webpage),
            'formats': formats,
            'thumbnail': video.get('yt_thumb_url') or self._html_search_meta(
                'twitter:image', webpage, default=None),
            'duration': video.get('duration_seconds'),
            'timestamp': video.get('created_date'),
            'location': video.get('venue'),
            'series': video.get('series_name'),
        }
```

File: yt_dlp/extractor/willow.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from ..compat import compat_str
```
Comment (pukkandan) [inline yt_dlp/extractor/willow.py:4]:
```suggestion
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
Please confirm that changes I made are fine

**Comment by aarubui** [comment]
I confirm.

**Pull request merged**